src/metabase/driver/util.clj

```diff
@@ -215,18 +215,19 @@
    critical metabase features that use this check."
   5000)
 
-(defn supports?
-  "A defensive wrapper around [[database-supports?]]. It adds logging and error handling to avoid crashing the app if this
-   method takes a long time to execute or throws an exception. This is useful because `supports?` is used in so many critical
-   places in the app, and we don't want a single driver to crash the app if it throws an exception, or delay the user if it
-   takes a long time to execute."
-  [driver feature database]
-  (try
-    (u/with-timeout supports?-timeout-ms
-      (driver/database-supports? driver feature database))
-    (catch Throwable e
-      (log/error e (u/format-color 'red "Failed to check feature '%s' for database '%s'" (name feature) (:name database)))
-      false)))
+(def ^{:arglists '([driver feature database])} supports?
+  "A defensive wrapper around [[database-supports?]]. It adds logging, caching, and error handling to avoid crashing the app
+   if this method takes a long time to execute or throws an exception. This is useful because `supports?` is used in so many
+   critical places in the app, and we don't want a single driver to crash the app if it throws an exception, or delay the user
+   if it takes a long time to execute."
+  (mdb/memoize-for-application-db ; memoize because this can be called in a tight loop, and will only change if the database changes versions
+   (fn [driver feature database]
```

Comment:
I've decided to memoize the function with the entire `database` instance as part of the key. It's likely that the return value of `driver/database-supports?` will only change on database upgrades, and therefore we wouldn't need to recompute the value of `driver.u/supports?` for changes like the name or description of the database. But it's better to be safe than sorry. We can't predict ahead of time what keys might be added to `database` instances that `driver/database-supports?` implementations might depend on.

Reply:
TIL about `:dbms_version`. How do we populate that?

Reply:
It's the first sync step:
https://github.com/metabase/metabase/blob/f7b065dc5af6a70b8a5c702b08d64e813aff41f6/src/metabase/sync/sync_metadata.clj#L50